toolkit/tools/scheduler/schedulerutils/depsolver.go

```diff
@@ -76,8 +100,12 @@ func LeafNodes(pkgGraph *pkggraph.PkgGraph, graphMutex *sync.RWMutex, goalNode *
 		// Ideally we will wait for the actual provider of the implicit node to be built and convert the implicit node to
 		// a normal node via InjectMissingImplicitProvides().
 		if !useCachedImplicit && pkgNode.Implicit && pkgNode.State == pkggraph.StateCached {
-			logger.Log.Debugf("Skipping cached implicit provide leaf node: %v", pkgNode)
-			return
+			if isToolchainPackage(pkgNode, toolchainPackages) {
```

Comment:
With this change we started to make a distinction between toolchain and non-toolchain implicit nodes and that distinction forces us to pass around the list of toolchain packages to this and other functions. Wouldn't it make sense to find toolchain nodes once when we build the graph and encode that knowledge in the node itself by creating a new type? Also, some of the checks above like `useCachedImplicit` and `pkgNode.State == pkggraph.StateCached` don't really apply to toolchain packages - they are always available in the cache. If they weren't, that would be a bug somewhere a lot earlier in the code, which might be another indication that toolchain and non-toolchain implicit nodes are different enough to give them separate types.